File: IT/Multithreading/Esercizi/C++/esercizio25.cpp
/*
file esercizio25.cpp
compilare con: g++ -std=c++11 esercizio25.cpp -lpthread -I.
oppure (c++20):
g++ -std=c++20 esercizio25.cpp -lpthread
*/

#include <random>
#include <thread>
#include <chrono>
#include <deque>
#if __cplusplus > 201703L
#include <semaphore>
using std::binary_semaphore;
#else
#include "semaphore.h"
using binary_semaphore = counting_semaphore<1>;
#endif

using std::random_device;
using std::default_random_engine;
using std::uniform_int_distribution;
using std::thread;
using std::chrono::duration;
using std::chrono::seconds;
using std::ref;
using std::deque;

void philosopher(const char* name, deque<binary_semaphore>& forks, int rIndex, int lIndex) {
	thread_local default_random_engine gen(random_device{}());

	while (true) {
		if (forks[rIndex].try_acquire()) {
			if (forks[lIndex].try_acquire()) {
				printf("%s prende le forchette %d e %d.\n", name, rIndex, lIndex);
				printf("%s sta mangiando.\n", name);
				// ogni filosofo impiega tra 4 e 8 secondi per mangiare
				uniform_int_distribution<int> dist(4, 8);
				seconds duration(dist(gen));
				std::this_thread::sleep_for(duration);

				printf("%s posa le forchette %d e %d.\n", name, rIndex, lIndex);
				forks[lIndex].release();
			}

			forks[rIndex].release();
		}

		// ogni filosofo pensa per 3/5 secondi
		uniform_int_distribution<int> dist(3, 5);
		seconds duration(dist(gen));
		std::this_thread::sleep_for(duration);
		printf("%s sta pensando...\n", name);
	}
}

int main()
{
	const int N_PHILOSOPHERS = 5;
	const char* names[] = {"Platone", "Pitagora", "Kant", "Leibniz", "Marx"};

	deque<binary_semaphore> forks;
	for (int i = 0; i < 5; i++) forks.emplace_back(1);

	thread philosophers[N_PHILOSOPHERS];

	for (int i = 0; i < N_PHILOSOPHERS; i++) {
		philosophers[i] = thread(philosopher, names[i], ref(forks), i, (i+1) % N_PHILOSOPHERS);
	}

	for (auto& thread : philosophers) {
		thread.join();
	}
}

File: IT/Multithreading/Esercizi/Java/Esercizio25.java
/* file Esercizio25.java */

import java.util.concurrent.ThreadLocalRandom;
import java.util.concurrent.Semaphore;

class Philosopher implements Runnable {
	private Semaphore [] forks;
	private int rIndex;
	private int lIndex;

	public Philosopher(Semaphore [] forks, int rIndex, int lIndex) {
		this.forks = forks;
		this.rIndex = rIndex;
		this.lIndex = lIndex;
	}

	@Override
	public void run() {
		Thread t = Thread.currentThread();
		while (true) {
			try {
				if (forks[rIndex].tryAcquire()) {
					if (forks[lIndex].tryAcquire()) {
						System.out.println(t.getName() + " prende le forchette " + rIndex + " e " + lIndex + ".");
						System.out.println(t.getName() + " sta mangiando.");

						// ogni filosofo impiega tra 4 e 8 secondi per mangiare
						int pause = ThreadLocalRandom.current().nextInt(4, 8+1);
						t.sleep(pause*1000);

						System.out.println(t.getName() + " posa le forchette " + rIndex + " e " + lIndex + ".");

						forks[lIndex].release();
					}

					forks[rIndex].release();
				}

				// ogni filosofo pensa per 3/5 minuti
				int pause = ThreadLocalRandom.current().nextInt(3, 5+1);
				t.sleep(pause*1000);
				System.out.println(t.getName() + " sta pensando...");
			}
			catch (InterruptedException e) {
				e.printStackTrace();
			}
		}
	}
}

public class Esercizio25 {
	public static void main(String [] args) {
		final int N_PHILOSOPHERS = 5;
		String [] names = {"Platone", "Pitagora", "Kant", "Leibniz", "Marx"};
		Semaphore [] forks = new Semaphore[N_PHILOSOPHERS];

		for (int i = 0; i < N_PHILOSOPHERS; i++) {
			forks[i] = new Semaphore(1);
		}

		for (int i = 0; i < N_PHILOSOPHERS; i++) {
			Thread philosopher = new Thread(new Philosopher(forks, i, (i+1) % N_PHILOSOPHERS), names[i]);
			philosopher.start();
		}
	}
}

